postgis/lwgeom_out_mvt.c

```diff
@@ -45,26 +45,76 @@ Datum ST_AsMVTGeom(PG_FUNCTION_ARGS)
 	elog(ERROR, "Missing libprotobuf-c");
 	PG_RETURN_NULL();
 #else
-	LWGEOM *lwgeom_in, *lwgeom_out;
+	GBOX *bounds = NULL;
+	int32_t extent = 0;
+	int32_t buffer = 0;
+	bool clip_geom = true;
 	GSERIALIZED *geom_in, *geom_out;
-	GBOX *bounds;
-	int extent, buffer;
-	bool clip_geom;
+	LWGEOM *lwgeom_in, *lwgeom_out;
+	uint8_t type = 0;
+
 	if (PG_ARGISNULL(0))
+	{
 		PG_RETURN_NULL();
-	geom_in = PG_GETARG_GSERIALIZED_P_COPY(0);
-	lwgeom_in = lwgeom_from_gserialized(geom_in);
+	}
+
 	if (PG_ARGISNULL(1))
+	{
 		elog(ERROR, "%s: parameter bounds cannot be null", __func__);
-	bounds = (GBOX *) PG_GETARG_POINTER(1);
+		PG_RETURN_NULL();
+	}
+	bounds = (GBOX *)PG_GETARG_POINTER(1);
+	if (bounds->xmax - bounds->xmin <= 0 || bounds->ymax - bounds->ymin <= 0)
+	{
+		elog(ERROR, "%s: bounds width or height cannot be 0", __func__);
+		PG_RETURN_NULL();
+	}
+
 	extent = PG_ARGISNULL(2) ? 4096 : PG_GETARG_INT32(2);
+	if (extent <= 0)
+	{
+		elog(ERROR, "%s: extent cannot be 0", __func__);
+		PG_RETURN_NULL();
+	}
+
 	buffer = PG_ARGISNULL(3) ? 256 : PG_GETARG_INT32(3);
 	clip_geom = PG_ARGISNULL(4) ? true : PG_GETARG_BOOL(4);
-	// NOTE: can both return in clone and in place modification so
-	// not known if lwgeom_in can be freed
+
+	geom_in = PG_GETARG_GSERIALIZED_P_COPY(0);
+	type = gserialized_get_type(geom_in);
+
+	/* If possible, peak into the bounding box before deserializating it to discard small geometries
+	 * We don't check COLLECTIONTYPE since that might be a collection of points */
+	if (type == LINETYPE || type == POLYGONTYPE || type == MULTILINETYPE || type == MULTIPOLYGONTYPE)
+	{
+		GBOX gserialized_box;
+		/* We only apply the optimization if the bounding box is available */
+		if (gserialized_read_gbox_p(geom_in, &gserialized_box) == LW_SUCCESS)
```

Comment:
there was a _peek function that also worked for two-point linestrings without box, maybe use it?

Reply:
I wanted to avoid repeating work in the common case (where this speedup isn't applied): in that case he bounding box would be calculated first for this check and later when deserializing the geometry.

Reply:
here, _peek is only used in trivial cases:

https://github.com/postgis/postgis/blob/d1e5a63aac0078b6699702d8758dc1f0c7714841/liblwgeom/g_serialized.c#L639

I suggest going first two if-branches, not all three. It feels it can be refactored so that _peek calls _read first instead of failing on present box, and here you just call a _peek.

Reply:
Box is not written out for everything that _peek has code for. 

https://github.com/postgis/postgis/blob/682db5db8a2113d41214458143a2f74125ad178e/liblwgeom/lwgeom.c#L1181